File: implementations/react-native/design-patterns-rn/src/patterns/behavioral/visitor/visitor-delivery.ts
import type { Book } from './element-book';
import type { Electronics } from './element-electronics';
import type { Grocery } from './element-grocery';
import type { Visitor } from './visitor';

export class ShippingEstimatorVisitor implements Visitor {
  shipping = 0;

  visitBook(e: Book) {
    this.shipping += 2 * e.qty;                      // light fee
  }

  visitElectronics(e: Electronics) {
    this.shipping += (e.fragile ? 10 : 6) * e.qty;   // fragile insurance
  }

  visitGrocery(e: Grocery) {
    this.shipping += (e.perishable ? 8 : 4) * e.qty; // cold chain
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/visitor/visitor_delivery.dart
import 'visitor.dart';
import 'element_book.dart';
import 'element_electronics.dart';
import 'element_grocery.dart';

class ShippingEstimatorVisitor implements Visitor {
  double shipping = 0;

  @override
  void visitBook(Book e) {
    shipping += 2 * e.qty; // light tariff
  }

  @override
  void visitElectronics(Electronics e) {
    shipping += (e.fragile ? 10 : 6) * e.qty; // fragile insured
  }

  @override
  void visitGrocery(Grocery e) {
    shipping += (e.perishable ? 8 : 4) * e.qty; // cold chain
  }
}
